File: Server_Root/Server/RoomServer/GameEvent/GameEvents/Position.cpp
#include "stdafx.h"
#include "Position.h"
#include "Position_generated.h"

namespace GameEvent
{
	Position::Position()
	{
	}

	Position::~Position()
	{
	}

	FBS::GameEventType Position::GetType()
	{
		return FBS::GameEventType::GameEventType_Position;
	}

	const char* Position::Serialize(int* pLength)
	{
		FBSData::Vector3 startPosition(m_vec3StartPosition.x(), m_vec3StartPosition.y(), m_vec3StartPosition.z());
		FBSData::Vector3 endPosition(m_vec3EndPosition.x(), m_vec3EndPosition.y(), m_vec3EndPosition.z());

		FBS::GameEvent::PositionBuilder data_builder(m_Builder);
		data_builder.add_EntityID(m_nEntityID);
		data_builder.add_StartTime(m_fStartTime);
		data_builder.add_EndTime(m_fEndTime);
		data_builder.add_StartPos(&startPosition);
		data_builder.add_EndPos(&endPosition);

		auto data = data_builder.Finish();

		m_Builder.Finish(data);

		*pLength = m_Builder.GetSize();

		return (char*)m_Builder.GetBufferPointer();
	}
}

File: Client_Root/Client/Assets/Scripts/GameEvent/GameEvents/Position.cs
using UnityEngine;
using FlatBuffers;

namespace GameEvent
{
    public class Position : IGameEvent
    {
        public int      m_nEntityID;
        public float    m_fStartTime;
        public float    m_fEndTime;
        public Vector3  m_vec3StartPosition;
        public Vector3  m_vec3EndPosition;

        public override FBS.GameEventType GetEventType()
        {
            return FBS.GameEventType.Position;
        }

        public override bool Deserialize(byte[] bytes)
        {
            var buf = new ByteBuffer(bytes);

            var data = FBS.GameEvent.Position.GetRootAsPosition(buf);

            m_nEntityID = data.EntityID;
            m_fStartTime = data.StartTime;
            m_fEndTime = data.EndTime;
            m_vec3StartPosition = new Vector3(data.StartPos.X, data.StartPos.Y, data.StartPos.Z);
            m_vec3EndPosition = new Vector3(data.EndPos.X, data.EndPos.Y, data.EndPos.Z);

            return true;
        }

        public override string ToString()
        {
            return string.Format("{0}, EntityID : {1}, StartTime : {2}, EndTime : {3}, StartPosition : {4}, EndPosition : {5}", base.ToString(), m_nEntityID, m_fStartTime, m_fEndTime, m_vec3StartPosition.ToString(), m_vec3EndPosition.ToString());
        }
    }
}
